File: java/L200/LeetCode0260.java
package L200;

public class LeetCode0260 {
    public int[] singleNumber(int[] nums) {
        int xor = 0;
        for (int num : nums) {
            xor ^= num;
        }
        int k = 0;
        for (int i = 31;i >= 0;--i) {
            if (((xor >>> i) & 1) == 1) {
                k = i;
                break;
            }
        }
        int[] ans = new int[2];
        for (int num : nums) {
            if (((num >>> k) & 1) == 1) {
                ans[0] ^= num;
            } else {
                ans[1] ^= num;
            }
        }
        return ans;
    }
}

File: cpp/L200/LeetCode0260.cc
#include <functional>
#include <limits>
#include <numeric>
#include <vector>
using std::vector;

class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int x{std::accumulate(nums.begin(), nums.end(), 0, std::bit_xor<int>())};
        x = x == std::numeric_limits<int>::min() ? std::numeric_limits<int>::min() : x & (-x);
        vector<int> ans{0, 0};
        for (int num : nums) {
            if ((num & x) != 0) {
                ans[0] = ans[0] ^ num;
            } else {
                ans[1] = ans[1] ^ num;
            }
        }
        return ans;
    }
};